libs/completer/completer.go

```diff
@@ -0,0 +1,66 @@
+package completer
+
+import (
+	"context"
+	"path"
+
+	"github.com/databricks/cli/libs/filer"
+	"github.com/spf13/cobra"
+)
+
+type completer struct {
+	ctx context.Context
+
+	// The filer to use for completing remote or local paths.
+	filer filer.Filer
+
+	// CompletePath will only return directories when onlyDirs is true.
+	onlyDirs bool
+}
+
+// General completer that takes a filer to complete remote paths when TAB-ing through a path.
+func New(ctx context.Context, filer filer.Filer, onlyDirs bool) *completer {
+	return &completer{ctx: ctx, filer: filer, onlyDirs: onlyDirs}
+}
+
+func (c *completer) CompletePath(p string) ([]string, string, cobra.ShellCompDirective, error) {
+	// If the user is TAB-ing their way through a path, the path in `toComplete`
+	// is valid and we should list nested directories.
+	// If the path in `toComplete` is incomplete, however,
+	// then we should list adjacent directories.
+	dirPath := p
+	completions, err := fetchCompletions(c, dirPath)
+	if completions == nil && err == nil {
+		dirPath = path.Dir(p)
```

Comment:
Nit: you can skip this if `dirPath == p`, i.e. at the root.

Reply:
Moved to using trailing slash for complete paths, so this is not needed